[AArch64] Use isKnownNonZero to optimize eligible compares to cmn and ccmn

AZero13 opened the pull request:
The problematic case for unsigned comparisons occurs only when the second argument is zero.

The problematic case for signed comparisons occurs only when the second argument is the signed minimum value.

We can use KnownBits to know when we don't have to worry about this.

Source: https://devblogs.microsoft.com/oldnewthing/20210607-00/?p=105288

AZero13 committed:
Pre-commit tests (NFC)

AtariDreams commented:
@arsenm What do you think of this?

AtariDreams commented:
@dtcxzyw does this have an adverse effect on compile time?

davemgreen commented on llvm/test/CodeGen/AArch64/cmp-chains.ll:
```diff
@@ -258,3 +258,34 @@ define i32 @neg_range_int(i32 %a, i32 %b, i32 %c) {
 }
 
+; (b > -3 || a < -(c | 1))
+define i32 @neg_range_int_cmn(i32 %a, i32 %b, i32 %c) {
```
I think you can add simpler test cases that do not involve ccmp.

Are you sure this is valid to transform?

AtariDreams replied:
c | 1 means it's never 0, so it should be valid.

davemgreen replied:
In the new code in this test:
```
define i32 @or_neg(i32 %x, i32 %y) {
; CHECK-LABEL: or_neg:
; CHECK:       // %bb.0:
; CHECK-NEXT:    orr w8, w0, #0x1
; CHECK-NEXT:    cmn w8, w1
; CHECK-NEXT:    cset w0, gt
; CHECK-NEXT:    ret
  %3 = or i32 %x, 1
  %4 = sub i32 0, %3
  %5 = icmp sgt i32 %4, %y
  %6 = zext i1 %5 to i32
  ret i32 %6
}
```
If, say, x==0, y==0 then is the result not backwards?

AtariDreams replied:
It is and I am investigating

AtariDreams replied:
> In the new code in this test:
> 
> ```
> define i32 @or_neg(i32 %x, i32 %y) {
> ; CHECK-LABEL: or_neg:
> ; CHECK:       // %bb.0:
> ; CHECK-NEXT:    orr w8, w0, #0x1
> ; CHECK-NEXT:    cmn w8, w1
> ; CHECK-NEXT:    cset w0, gt
> ; CHECK-NEXT:    ret
>   %3 = or i32 %x, 1
>   %4 = sub i32 0, %3
>   %5 = icmp sgt i32 %4, %y
>   %6 = zext i1 %5 to i32
>   ret i32 %6
> }
> ```
> 
> If, say, x==0, y==0 then is the result not backwards?

Fixed

davemgreen replied:
I think this case (the ccmn one) is still incorrect, unfortunately.

AtariDreams replied:
> I think this case (the ccmn one) is still incorrect, unfortunately.

Fixed!

AtariDreams commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -26509,1968 +26538,4 @@ Value *AArch64TargetLowering::getSDagStackGuard(const Module &M) const {
   return TargetLowering::getSDagStackGuard(M);
 }
 
-Function *AArch64TargetLowering::getSSPStackGuardCheck(const Module &M) const {
```
I have no idea how my editor just removed this code. I didn't intend on that.

AtariDreams commented:
How are these tests failing? They pass on my machine just fine!

AtariDreams commented:
@arsenm @davemgreen Thoughts?

davemgreen reviewed (commented):
This is quite complex now, and the ccmn case is apparently still not happy. It would be useful to add more test cases for various conditions. or_neg looks good, but we should have test for all the cases that wouldn't be.

AtariDreams commented:
@davemgreen I am going to split this patch out. So this should be correct now.

AtariDreams commented:
@davemgreen This should be ready now!

AtariDreams commented:
@jonathandavies-arm What are your thoughts on this?

AtariDreams commented:
@davemgreen This should be ready now!

davemgreen reviewed (commented):
I think this looks OK now, and the test all look OK. Do you mind if we wait until after the release branch to commit, in case there is something still wrong?

It would be good to have test cases for lots of various cases too, even if they don't trigger - different signed conditions, cmn on the LHS/RHS, ccmn on the LHS/RHS etc. Thanks

AtariDreams commented:
> I think this looks OK now, and the test all look OK. Do you mind if we wait until after the release branch to commit, in case there is something still wrong?
> 
> 
> 
> It would be good to have test cases for lots of various cases too, even if they don't trigger - different signed conditions, cmn on the LHS/RHS, ccmn on the LHS/RHS etc. Thanks

If something goes wrong, we can revert it. However, I've tested and tested this and found no issue.

AtariDreams commented:
I'm working on an expansion to this but that will sadly have to wait until after the release branch, but this is the part that has been thoroughly tested

arsenm commented:
> If something goes wrong, we can revert it. However, I've tested and tested this and found no issue.

This isn't really the point of adding tests.

davemgreen commented:
OK - I was just being careful as we are so close to the release and some of the issues can be subtle. I gave it a more thorough test and everything seems OK.

By "tested this and found no issue", does that include running a bootstrap and the llvm-test-suite? If so and you add a some extra test coverage then this LGTM.

AtariDreams commented:
> OK - I was just being careful as we are so close to the release and some of the issues can be subtle. I gave it a more thorough test and everything seems OK.
> 
> 
> 
> By "tested this and found no issue", does that include running a bootstrap and the llvm-test-suite? If so and you add a some extra test coverage then this LGTM.

Yes I ran the entire test suite.

davemgreen commented:
OK. If that has succeeded and a bootstrap also does OK, then if you can add some more tests for various ccmn cases we can get this in.

AZero13 committed:
[AArch64] Use isKnownNonZero to optimize eligible compares to cmn and ccmn

The problematic case for unsigned comparisons occurs only when the second argument is zero, and in signed cases when the second argument is the minimum possible signed integer. If we can prove the register value be those, it is safe to fold into CMN and CCMN.

Source: https://devblogs.microsoft.com/oldnewthing/20210607-00/?p=105288

davemgreen reviewed (commented):
This case is now transforming, but I don't believe that is valid:
```
define i32 @and_ult_eq_s0s1_or(i32 %s0a, i32 %s1a, i32 %s2, i32 %s3) {
; CHECK-LABEL: and_ult_eq_s0s1_or:
; CHECK:       // %bb.0: // %entry
; CHECK-NEXT:    orr w8, w1, #0x1
; CHECK-NEXT:    orr w9, w0, #0x1
; CHECK-NEXT:    cmp w2, w3
; CHECK-NEXT:    ccmn w8, w9, #2, eq
; CHECK-NEXT:    mov w8, #20 // =0x14
; CHECK-NEXT:    mov w9, #10 // =0xa
; CHECK-NEXT:    csel w0, w9, w8, lo
; CHECK-NEXT:    ret
entry:
  %a1 = or i32 %s0a, 1
  %s1 = or i32 %s1a, 1
  %s0 = sub i32 0, %a1
  %c0 = icmp ult i32 %s0, %s1
  %c1 = icmp eq i32 %s2, %s3
  %a = and i1 %c0, %c1
  %r = select i1 %a, i32 10, i32 20
  ret i32 %r
}
```

AtariDreams commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -3566,13 +3574,21 @@ static SDValue emitConditionalComparison(SDValue LHS, SDValue RHS,
+  } else if (isCMN(RHS, CC, DAG)) {
+    Opcode = AArch64ISD::CCMN;
+    RHS = RHS.getOperand(1);
+  } else if (isCMN(LHS, CC, DAG) &&
+             (isIntEqualitySetCC(CC) ||
+              (isUnsignedIntSetCC(CC) && DAG.isKnownNeverZero(RHS)) ||
+              (isSignedIntSetCC(CC) && cannotBeIntMin(RHS, DAG)))) {
```
Yes I do!

Let me show you!

AtariDreams replied:
https://alive2.llvm.org/ce/z/8YTkie

AtariDreams replied:
Actually if you want I can remove that part, but I have proof it works

AtariDreams replied:
You know, I will revert this part only because it is risky keeping it is even though it is proven and tested

AtariDreams replied:
Proof: 

It is known that if a > b, then b < a,

it is known that if b < a, then -b > -a

For the degenerative cases of INT_MIN and 0 for signed and unsigned comparisons, this is where knownbits steps in: https://alive2.llvm.org/ce/z/8YTkie

AtariDreams replied:
I reverted the changed part so we can approve this. I will split the enhancement for another PR.

davemgreen replied:
I think the and_ult_eq_s0s1_or case was mis-compiling because of
```
    // Swap LHS and RHS if it wasn't an equality comparison
    // So we don't have to worry about changing the CC
    // a < b -> -b < -a
    std::swap(LHS, RHS);
```
The condition needed to be updated too.

AtariDreams replied:
I swapped it so that the condition didn't need to be updated.

AtariDreams replied:
> I think the and_ult_eq_s0s1_or case was mis-compiling because of
> 
> ```
>     // Swap LHS and RHS if it wasn't an equality comparison
>     // So we don't have to worry about changing the CC
>     // a < b -> -b < -a
>     std::swap(LHS, RHS);
> ```
> 
> The condition needed to be updated too.

The condition does not need to be updated because:

a < b -> -a > -b -> -b < -a

AtariDreams replied:
Negation of both sides flips the condition and the swap flips it back. I did this because the CC isn't used in this method except for testing for eligibility, and this is much more feasible than refactoring so that we recompute the condition codes and arm condition codes.

AtariDreams commented:
> This case is now transforming, but I don't believe that is valid:
> 
> ```
> define i32 @and_ult_eq_s0s1_or(i32 %s0a, i32 %s1a, i32 %s2, i32 %s3) {
> ; CHECK-LABEL: and_ult_eq_s0s1_or:
> ; CHECK:       // %bb.0: // %entry
> ; CHECK-NEXT:    orr w8, w1, #0x1
> ; CHECK-NEXT:    orr w9, w0, #0x1
> ; CHECK-NEXT:    cmp w2, w3
> ; CHECK-NEXT:    ccmn w8, w9, #2, eq
> ; CHECK-NEXT:    mov w8, #20 // =0x14
> ; CHECK-NEXT:    mov w9, #10 // =0xa
> ; CHECK-NEXT:    csel w0, w9, w8, lo
> ; CHECK-NEXT:    ret
> entry:
>   %a1 = or i32 %s0a, 1
>   %s1 = or i32 %s1a, 1
>   %s0 = sub i32 0, %a1
>   %c0 = icmp ult i32 %s0, %s1
>   %c1 = icmp eq i32 %s2, %s3
>   %a = and i1 %c0, %c1
>   %r = select i1 %a, i32 10, i32 20
>   ret i32 %r
> }
> ```

It is correct: ccmn w8, w9 can be swapped around while testing for equality: if w8 == -w9, then -w8 == w9.

AtariDreams commented:
@davemgreen I have reverted this back because otherwise I'm gonna have to do another bootstrap test, and I think that change can be accepted after this one instead of part of this PR.

So this should be good to go!

davemgreen reviewed (approved):
I'm happy with this version. LGTM
